File: doacao_app_java/src/main/java/DAO/AnimalDAO.java
package DAO;

import schemas.Animal;
import repository.MYSQLConnection;
import exceptions.CustomException;
import schemas.Protetor;

import java.sql.*;
import java.util.ArrayList;
import java.util.List;

public class AnimalDAO {
    private Connection connection;

    public AnimalDAO() throws CustomException {
        try {
            this.connection = MYSQLConnection.getConnection();
        } catch (Exception e) {
            throw new CustomException("Erro ao conectar ao banco de dados", e);
        }
    }

    // CREATE: Inserir um novo Animal
    public void create(Animal animal) throws CustomException {
        String sql = "INSERT INTO Animal (especie, raca, temperamento, historicoSaude, nome, descricao, esEspecial, idade, sexo, status, id_protetor) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setString(1, animal.getEspecie());
            stmt.setString(2, animal.getRaca());
            stmt.setString(3, animal.getTemperamento());
            stmt.setString(4, animal.getHistoricoSaude());
            stmt.setString(5, animal.getNome());
            stmt.setString(6, animal.getDescricao());
            stmt.setBoolean(7, animal.getEsEspecial());
            stmt.setInt(8, animal.getIdade());
            stmt.setString(9, String.valueOf(animal.getSexo()));
            stmt.setString(10, animal.getStatus());
            stmt.setInt(11, animal.getProtetor().getId());
            stmt.executeUpdate();
            System.out.println("Animal criado com sucesso!");
        } catch (SQLException e) {
            throw new CustomException("Erro ao criar Animal", e);
        }
    }

    // READ: Buscar um Animal por ID
    public Animal read(int idAnimal) throws CustomException {
        String sql = "SELECT * FROM Animal WHERE idAnimal = ?";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setInt(1, idAnimal);
            ResultSet rs = stmt.executeQuery();
            if (rs.next()) {
                Animal animal = new Animal();
                animal.setId(rs.getInt("idAnimal"));
                animal.setEspecie(rs.getString("especie"));
                animal.setRaca(rs.getString("raca"));
                animal.setTemperamento(rs.getString("temperamento"));
                animal.setHistoricoSaude(rs.getString("historicoSaude"));
                animal.setNome(rs.getString("nome"));
                animal.setDescricao(rs.getString("descricao"));
                animal.setEsEspecial(rs.getBoolean("esEspecial"));
                animal.setIdade(rs.getInt("idade"));
                animal.setSexo(rs.getString("sexo").charAt(0));
                animal.setStatus(rs.getString("status"));

                // Buscar o objeto agregado Protetor
                ProtetorDAO protetordao = new ProtetorDAO();
                Protetor protetor = protetordao.read(rs.getInt("id_protetor"));
                animal.setProtetor(protetor);

                return animal;
            }
        } catch (SQLException e) {
            throw new CustomException("Erro ao buscar Animal", e);
        }
        return null;
    }

    // Listar todos os Animals
    public List<Animal> readAll() throws CustomException {
        List<Animal> animais = new ArrayList<>();
        String sql = "SELECT * FROM Animal";
        try (Statement stmt = connection.createStatement();
             ResultSet rs = stmt.executeQuery(sql)) {
            while (rs.next()) {
                Animal animal = new Animal();
                animal.setId(rs.getInt("idAnimal"));
                animal.setEspecie(rs.getString("especie"));
                animal.setRaca(rs.getString("raca"));
                animal.setTemperamento(rs.getString("temperamento"));
                animal.setHistoricoSaude(rs.getString("historicoSaude"));
                animal.setNome(rs.getString("nome"));
                animal.setDescricao(rs.getString("descricao"));
                animal.setEsEspecial(rs.getBoolean("esEspecial"));
                animal.setIdade(rs.getInt("idade"));
                animal.setSexo(rs.getString("sexo").charAt(0));
                animal.setStatus(rs.getString("status"));


                ProtetorDAO protetordao = new ProtetorDAO();
                Protetor protetor = protetordao.read(rs.getInt("id_protetor"));
                animal.setProtetor(protetor);

                animais.add(animal);
            }
        } catch (SQLException e) {
            throw new CustomException("Erro ao listar Animais", e);
        }
        return animais;
    }

    public void update(Animal animal) throws CustomException {
        String sql = "UPDATE Animal SET especie = ?, raca = ?, temperamento = ?, historicoSaude = ?, nome = ?, descricao = ?, esEspecial = ?, idade = ?, sexo = ?, status = ?, id_protetor = ? WHERE idAnimal = ?";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setString(1, animal.getEspecie());
            stmt.setString(2, animal.getRaca());
            stmt.setString(3, animal.getTemperamento());
            stmt.setString(4, animal.getHistoricoSaude());
            stmt.setString(5, animal.getNome());
            stmt.setString(6, animal.getDescricao());
            stmt.setBoolean(7, animal.getEsEspecial());
            stmt.setInt(8, animal.getIdade());
            stmt.setString(9, String.valueOf(animal.getSexo()));
            stmt.setString(10, animal.getStatus());
            stmt.setInt(11, animal.getProtetor().getId());
            stmt.setInt(12, animal.getId());
            stmt.executeUpdate();
            System.out.println("Animal atualizado com sucesso!");
        } catch (SQLException e) {
            throw new CustomException("Erro ao atualizar Animal", e);
        }

    }

    // DELETE: Deletar um Animal por ID
    public void delete(int id) throws CustomException {
        String sql = "DELETE FROM Animal WHERE idAnimal = ?";
        try (PreparedStatement stmt = connection.prepareStatement(sql)) {
            stmt.setInt(1, id);
            stmt.executeUpdate();
            System.out.println("Animal deletado com sucesso!");
        } catch (SQLException e) {
            throw new CustomException("Erro ao deletar Animal", e);
        }
    }
}

File: doacao_animal_python/DAO/animal_dao.py
from repository.mysql_connection import MYSQLConnection
from exceptions.custom_exception import CustomException
from schemas.animal import Animal


class AnimalDAO:
    @staticmethod
    def create(animal: Animal) -> None:
        conn = MYSQLConnection.get_connection()
        cursor = conn.cursor()
        sql = """
        INSERT INTO Animal (especie, raca, temperamento, historicoSaude, nome, descricao, esEspecial, idade, sexo, status, id_protetor)
        VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        """
        try:
            cursor.execute(sql, (
                animal.especie, animal.raca, animal.temperamento, animal.historicoSaude,
                animal.nome, animal.descricao, animal.esEspecial, animal.idade,
                animal.sexo, animal.status, animal.protetor.id
            ))
            conn.commit()
            print("Animal criado com sucesso!")
        except Exception as e:
            raise CustomException(f"Erro ao criar Animal: {e}")
        finally:
            cursor.close()

    @staticmethod
    def read(id: int) -> Animal:
        conn = MYSQLConnection.get_connection()
        cursor = conn.cursor(dictionary=True)
        sql = "SELECT * FROM Animal WHERE idAnimal = %s"
        try:
            cursor.execute(sql, (id,))
            row = cursor.fetchone()
            if row:
                # Buscar Protetor
                from DAO.protetor_dao import ProtetorDAO
                protetor = ProtetorDAO.read(row['id_protetor'])
                # Buscar Fotos
                from DAO.foto_animal_dao import FotoAnimalDAO
                fotos = FotoAnimalDAO.read_by_animal(id)
                return Animal(
                    idAnimal=row['idAnimal'],
                    especie=row['especie'],
                    raca=row['raca'],
                    temperamento=row['temperamento'],
                    historicoSaude=row['historicoSaude'],
                    nome=row['nome'],
                    descricao=row['descricao'],
                    esEspecial=row['esEspecial'],
                    idade=row['idade'],
                    sexo=row['sexo'],
                    status=row['status'],
                    protetor=protetor,
                    foto_animal=fotos
                )
            return None
        except Exception as e:
            raise CustomException(f"Erro ao buscar Animal: {e}")
        finally:
            cursor.close()

    @staticmethod
    def readAll() -> list[Animal]:
        conn = MYSQLConnection.get_connection()
        cursor = conn.cursor(dictionary=True)
        sql = "SELECT * FROM Animal"
        try:
            cursor.execute(sql)
            rows = cursor.fetchall()
            animais = []
            for row in rows:
                # Buscar Protetor e Fotos para cada
                from DAO.protetor_dao import ProtetorDAO
                protetor = ProtetorDAO.read(row['id_protetor'])
                from DAO.foto_animal_dao import FotoAnimalDAO
                fotos = FotoAnimalDAO.read_by_animal(row['idAnimal'])
                animais.append(Animal(
                    idAnimal=row['idAnimal'], especie=row['especie'], raca=row['raca'],
                    temperamento=row['temperamento'], historicoSaude=row['historicoSaude'],
                    nome=row['nome'], descricao=row['descricao'], esEspecial=row['esEspecial'],
                    idade=row['idade'], sexo=row['sexo'], status=row['status'], protetor=protetor, foto_animal=fotos
                ))
            return animais
        except Exception as e:
            raise CustomException(f"Erro ao listar Animais: {e}")
        finally:
            cursor.close()

    @staticmethod
    def update(animal: Animal) -> None:
        conn = MYSQLConnection.get_connection()
        cursor = conn.cursor()
        sql = """
        UPDATE Animal SET especie=%s, raca=%s, temperamento=%s, historicoSaude=%s, nome=%s, descricao=%s, esEspecial=%s, idade=%s, sexo=%s, status=%s, id_protetor=%s
        WHERE idAnimal=%s
        """
        try:
            cursor.execute(sql, (
                animal.especie, animal.raca, animal.temperamento, animal.historicoSaude,
                animal.nome, animal.descricao, animal.esEspecial, animal.idade,
                animal.sexo, animal.status, animal.protetor.id, animal.idAnimal
            ))
            conn.commit()
            print("Animal atualizado com sucesso!")
        except Exception as e:
            raise CustomException(f"Erro ao atualizar Animal: {e}")
        finally:
            cursor.close()

    @staticmethod
    def delete(id: int) -> None:
        conn = MYSQLConnection.get_connection()
        cursor = conn.cursor()
        sql = "DELETE FROM Animal WHERE idAnimal = %s"
        try:
            cursor.execute(sql, (id,))
            conn.commit()
            print("Animal deletado com sucesso!")
        except Exception as e:
            raise CustomException(f"Erro ao deletar Animal: {e}")
        finally:
            cursor.close()